### PR TITLE
Release closed streams capacity back to connection

### PR DESCRIPTION
**Pull request by seanmonstar**
Previously, any streams that were dropped or closed while not having
consumed the inflight received window capacity would simply leak that
capacity for the connection. This could easily happen if a `RecvStream`
were dropped before fully consuming the data, and therefore a user would
have no idea how much capacity to release in the first place. This
resulted in stalled connections that would never have capacity again.

### CODE REVIEW

File: src/proto/streams/recv.rs [hunk 1]
```diff
@@ -544,6 +567,12 @@ impl Recv {
             }
         }
 
+        // Update stream level flow control
+        stream.recv_flow.send_data(sz);
+
+        // Track the data as in-flight
+        stream.in_flight_recv_data += sz;
+
```
Comment (olix0r) [inline src/proto/streams/recv.rs]: just for my own understanding, was this change required subtly? It seems that the effect is just that we don't update the stream state when the stream protocol errors, which seems fine. Anything else here?

Comment (seanmonstar) [reply src/proto/streams/recv.rs]: Yes it was. Without the change, another test was failing (yay having tests!). Before the change, the size was added to the stream's `in_flight_recv_data`, but then the returned error already returns the window size for this data frame to the connection: https://github.com/carllerche/h2/blob/master/src/proto/streams/streams.rs#L240-L247

In this PR, with the `in_flight_recv_data` being returned to the connection when `ref_count` becomes zero meant there was window capacity returned twice, causing an underflow.
